File: fiddling/static/systems/systems.js
import { garbage_filter } from '../utils.js'

export class System {
    constructor(){
	this.components = []
	this.components_dict = null
    }
    get_component(type, entity){
	var component = new this.components_dict[type](entity)
	this.components.push(component)
	return component
    }
    garbage_collect(){
	garbage_filter(this.components, component => component.entity.lifetime <= 0)
    }
}
	

File: fiddling/server/systems/systems.py
from server.utils import garbage_filter


class System:
    def __init__(self):
        self.components = []
        self.components_dict = None

    def get_component(self, entity, type, *args, **kwargs):
        component = self.components_dict[type](entity, *args, **kwargs)
        self.components.append(component)
        return component

    def garbage_collect(self):
        garbage_filter(self.components, lambda c: c.entity.to_delete)

    def update(self, world, dt):
        self.garbage_collect()
        for component in self.components:
            component.update(dt)
